File: frontend/lib/features/card/domain/heart_card.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'heart_card.freezed.dart';
part 'heart_card.g.dart';

/// 心动卡片模型
@freezed
sealed class HeartCard with _$HeartCard {
  const factory HeartCard({
    @Default('') String imageUrl,
    @Default(0.0) double matchScore,
    required int id,
    required int userId,
    required String scene,
    @Default('') String sceneLabel,
    @Default('') String location,
    required String occurredAt,
    @Default('') String topColor,
    @Default('') String pantsColor,
    @Default(0) int glasses,    // 0-未知 1-有 2-无
    @Default('') String hairstyle,
    @Default(0) int hasBag,     // 0-未知 1-有 2-无
    @Default('') String shoeColor,
    @Default('') String description,
    @Default(1) int status,     // 1-有效 2-已匹配 3-已过期
    required String expireAt,
    required String createdAt,
  }) = _HeartCard;

  factory HeartCard.fromJson(Map<String, dynamic> json) =>
      _$HeartCardFromJson(json);

  const HeartCard._();

  // === 计算属性 ===

  /// 场景中文映射
  static const _sceneLabels = {
    'subway':  '🚇 地铁',
    'library': '📚 图书馆',
    'cafe':    '☕ 咖啡店',
    'campus':  '🏫 校园',
    'other':   '📍 其他',
  };

  String get sceneText => _sceneLabels[scene] ?? scene;

  /// 剩余分钟数
  int get remainingMinutes {
    final expire = DateTime.tryParse(expireAt);
    if (expire == null) return 0;
    return expire.difference(DateTime.now()).inMinutes.clamp(0, 30);
  }

  /// 是否即将过期（< 5 分钟）
  bool get isExpiringSoon => remainingMinutes < 5;

  /// 生成特征标签列表
  List<String> get featureTags {
    final tags = <String>[];
    if (topColor.isNotEmpty) tags.add('👔 $topColor');
    if (pantsColor.isNotEmpty) tags.add('👖 $pantsColor');
    if (hairstyle.isNotEmpty) tags.add('💇 $hairstyle');
    if (shoeColor.isNotEmpty) tags.add('👟 $shoeColor');
    if (glasses == 1) tags.add('👓 戴眼镜');
    if (hasBag == 1) tags.add('🎒 背包');
    return tags;
  }
}

File: backend/src/main/java/com/glance/model/entity/HeartCard.java
package com.glance.model.entity;

import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;

@Entity
@Table(name = "heart_card")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class HeartCard {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "user_id", nullable = false)
    private Long userId;

    @Column(nullable = false, length = 50)
    private String scene; // subway/library/cafe/other

    @Column(name = "scene_label", length = 100)
    @Builder.Default
    private String sceneLabel = "";

    @Column(length = 200)
    @Builder.Default
    private String location = "";

    @Column(name = "occurred_at", nullable = false)
    private LocalDateTime occurredAt;

    @Column(name = "top_color", length = 20)
    @Builder.Default
    private String topColor = "";

    @Column(name = "pants_color", length = 20)
    @Builder.Default
    private String pantsColor = "";

    @Column(nullable = false)
    @Builder.Default
    private Integer glasses = 0; // 0-未知 1-有 2-无

    @Column(length = 30)
    @Builder.Default
    private String hairstyle = "";

    @Column(name = "has_bag")
    @Builder.Default
    private Integer hasBag = 0; // 0-未知 1-有 2-无

    @Column(name = "shoe_color", length = 20)
    @Builder.Default
    private String shoeColor = "";

    @Column(length = 300)
    @Builder.Default
    private String description = "";

    @Column(nullable = false)
    @Builder.Default
    private Integer status = 1; // 1-有效 2-已匹配 3-已过期

    @Column(name = "expire_at", nullable = false)
    private LocalDateTime expireAt;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @PrePersist
    protected void onCreate() {
        createdAt = LocalDateTime.now();
        if (expireAt == null) {
            expireAt = createdAt.plusMinutes(30);
        }
    }
}
